Allow extension services to be discovered

paul-rogers opened the pull request:
### Description

Druid provides extensive extension support. Extension can define new services, but those services cannot yet be discovered. This PR ensures that they operate like native services. See [this ticket](https://github.com/apache/druid/issues/12218) for details.

The current code has a static list of node roles. This PR moves the list into a Guice multi-binder so that extensions can add to the list.

Next, the SQL system servers table code is revised to use the Guice-provided list of node roles rather than the hard-coded list.

Then, the `/druid/coordinator/v1/cluster` is added to include extension roles after the Druid-defined roles. The code imposed a specific order on the roles; those rules are preserved.

Finally, a new endpoint `/druid/router/v1/cluster` is added. The logic here is that clients start with the router endpoint. To get the list of services, they must first get the list of services to get the coordinator. But, they can't get the list of services without first haven gotten that list. To avoid this Catch-22, the new endpoint provides the list of services from the router itself.

Of course, the SQL servers system table also provides the list of services. The Catch-22 in this case is that if the cluster is broken, SQL is unavailable to help figure out which services are down. Having the native endpoint provides a reliable fallback: as long as the Router and ZK are up, we can learn about other services and see what is missing.

The PR includes some refactoring to make bits of role-related code usable in multiple contexts. Most of the code is not unit testable (we can't run servers in unit tests), but where it is testable, tests are added or modified.

<hr>

##### Key changed/added classes in this PR

 * `NodeRoles` - A Guice multi-binder to hold the list of Druid- and extension-defined node roles.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [X] been tested in a test Druid cluster, in the context of the actual extension service which this PR supports.

paul-rogers commented:
Tests mostly pass. There is one IT failure, but it looks like a flake rather than an actual failure. Would some committer please rerun that one test?

gianm commented on server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java:
```diff
@@ -0,0 +1,160 @@
+ * is fine because the actions don't occur frequently, nor do the
+ * actions take much time.
+ */
+public class LocalDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider implements DruidNodeAnnouncer
```
I don't see a way to activate this class outside of unit tests. Is it just meant for unit testing? If so, please call that out in the javadocs, and consider moving it to test scope. Or, if it's meant to be something we'll one day use outside of unit tests, please call that out in the javadocs, along with a reason as to why it isn't exposed yet.

gianm commented on server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java:
```diff
@@ -0,0 +1,160 @@
+    private void register(DiscoveryDruidNode node)
+    {
+      List<DruidNodeDiscovery.Listener> targets;
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
```
Seems like `nodes` is not properly synchronized; here the monitor is `LocalDruidNodeDiscoveryProvider.this` but in other methods it's `RoleEntry.this` by way of the method-level `synchronized` keyword.

gianm replied:
Suggestion: add errorprone's `@GuardedBy` annotation too, which will make the IDE complain at you if something is used without the proper synchronization.

gianm commented on server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java:
```diff
@@ -0,0 +1,160 @@
+  {
+    @SuppressWarnings("unused")
+    private final NodeRole role;
+    private final Map<DruidNode, DiscoveryDruidNode> nodes = new HashMap<>();
```
I'm not 100% sure if this is valid. I think it's possible for there to be more than one DiscoveryDruidNode per DruidNode, if a node has more than one NodeRole. Although I haven't dug into this to be sure.

gianm replied:
Ah I believe you're right. The structure in this file is NodeRole -> DruidNode -> DruidNodeDiscovery, which does match how the ZK one works. When I was looking at it earlier, I missed that this map here is scoped to a single NodeRole.

gianm commented on server/src/main/java/org/apache/druid/discovery/NodeRoles.java:
```diff
@@ -0,0 +1,101 @@
+   */
+  public static void addRole(Binder binder, NodeRole role)
+  {
+    LOG.debug("Adding node role: " + role.getJsonName());
```
Minor: it's a bit more conventional, Druid-wise, to do `log.debug("Adding node role: %s", role.getJsonName())`.

gianm commented on server/src/main/java/org/apache/druid/server/Node.java:
```diff
@@ -0,0 +1,99 @@
+import com.google.common.base.Objects;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class Node
```
This looks a lot like a cut-down DruidNode. To avoid confusion, the javadocs should describe the difference between this and DruidNode, and when we use one vs. the other.

gianm commented on server/src/test/java/org/apache/druid/server/NodeTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNull;
+
+public class NodeTest
```
Would also be good to have an equals verification test here. Like this:

```java
  @Test
  public void testEqualsAndHashCode()
  {
    EqualsVerifier.forClass(Node.class)
        .usingGetClass()
        .verify();
  }
```

paul-rogers replied:
And, the class is now just a facade on `DruidNode`, so we defer to the `equals()` and `hashCode()` methods on that class. (Though, we never actually use this class as a key: only a value.)

gianm reviewed (commented):
The concept looks good to me, but please let me know what you think of the line-by-line comments.

And, on testing: the integration test ITHighAvailabilityTest has a case that exercises discovery of an extended node role (CliCustomNodeRole) via DruidNodeDiscoveryProvider. If you add a case that verifies this through the new router API and the `sys.servers` table, that would constitute end-to-end testing of the new functionality. Which would be great.

paul-rogers commented:
Applied requested changes. Added integration tests. These are quite hard to test, so let's wait for the build to tell us what fixes may be needed before doing another review.

gianm commented on server/src/main/java/org/apache/druid/initialization/Initialization.java:
```diff
@@ -420,7 +421,6 @@ public static Injector makeInjectorWithModules(
         new IndexingServiceDiscoveryModule(),
         new CoordinatorDiscoveryModule(),
         new LocalDataStorageDruidModule(),
-        new TombstoneDataStorageModule(),
```
Is this removal intentional?

paul-rogers replied:
Good catch! Rebase error.

gianm commented on server/src/main/java/org/apache/druid/server/Node.java:
```diff
@@ -0,0 +1,92 @@
+ * This class exists only to support two REST endpoints. It is similar to a
+ * DruidNode, but serializes to the specific form expected by those REST
+ * endpoints. This version omits {@code bindOnHost}, uses the name
+ * {@code `service} where DruidNode uses {@code serviceName}, and
```
Stray backtick.

gianm commented on server/src/main/java/org/apache/druid/server/Node.java:
```diff
@@ -0,0 +1,92 @@
+import com.google.common.base.Objects;
+
+/**
+ * This class exists only to support two REST endpoints. It is similar to a
```
Which two?

paul-rogers commented on processing/src/main/java/org/apache/druid/query/metadata/metadata/NoneColumnIncluderator.java:
```diff
@@ -21,6 +21,8 @@
 /**
  */
+// Needed for IntelliJ checks
+@SuppressWarnings("unused")
```
Turned out not to be needed. Removed.

paul-rogers commented:
Having fun with ITs. In the previous commit, the one I added passed, multiple others failed. When restarted, the one I added failed, all others passed. Trying to find the cause of the intermittent failure. As it turns out, our auto-retry loop does not seem to provide the actual TestNG error so I'm flying blind. Will add logging and try the whole build *again*.

paul-rogers commented:
Converting to draft. Hard to debug the IT failures. Will return later when the ITs are more usable.

paul-rogers commented:
Rebased on latest master.

gianm commented on server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java:
```diff
@@ -0,0 +1,164 @@
+      listener.nodeViewInitialized();
+    }
+
+    private synchronized boolean contains(DruidNode node)
```
Does this method need to be synchronized on both `RoleEntry.this` and `LocalDruidNodeDiscoveryProvider.this`?

paul-rogers replied:
Good catch. No, it doesn't. The `LocalDruidNodeDiscoveryProvider` is the sync point. This isn't going to be a high-frequency API, so locking the main object is fine.

gianm commented:
Just wanted to make sure you saw this comment too, on the docs: https://github.com/apache/druid/pull/12222#discussion_r896236263

paul-rogers commented:
Failed in an ARM 64 supervisor test which appears to be [flaky](https://github.com/apache/druid/issues/12649).

paul-rogers commented on docs/operations/api-reference.md:
```diff
@@ -935,6 +935,12 @@ Returns segment information lists including server locations for the given query
 
 #### GET
 
+* `/druid/v2/router/cluster`
```
Fixed.

paul-rogers commented:
BTW: the change in this PR should be tested in an IT. At present, adding such an IT is quite a chore. Waiting for [the IT revision PR](https://github.com/apache/druid/pull/12368) to pass tests and merge, then it will be easy to add the required test.

rohangarg commented on server/src/main/java/org/apache/druid/discovery/NodeRoles.java:
```diff
@@ -0,0 +1,101 @@
+    return new HashSet<>(Arrays.asList(NodeRole.values()));
+  }
+
+  public static void addKnownRoles(Binder binder)
```
nit: `bindKnownRoles` could be used instead of `addKnownRoles`

rohangarg commented on server/src/main/java/org/apache/druid/discovery/NodeRoles.java:
```diff
@@ -0,0 +1,101 @@
+      boolean full)
+  {
+    if (full) {
+      return (Collection<Object>) (Collection<?>) getDiscoveryNodesForRole(provider, nodeRole);
```
seems like these could be removed now since the methods return collections themselves - also the suppression

rohangarg reviewed (approved):
Thanks for the changes! LGTM % minor comments

paul-rogers commented on server/src/main/java/org/apache/druid/initialization/Initialization.java:
```diff
@@ -518,7 +519,13 @@ private boolean shouldLoadOnCurrentNodeType(Object object)
-      return rolesPredicate.stream().anyMatch(nodeRoles::contains);
+      boolean shouldLoad = rolesPredicate.stream().anyMatch(nodeRoles::contains);
+      if (!shouldLoad) {
+        log.debug(
```
Changed. As it turns out, it started at info and someone suggested it be moved to debug. I added it because, like you, I had a case in which an extension was not being loaded and needed to know why.

paul-rogers commented:
No good deed goes unpunished: the changes from review comments broke something that shows up only in ITs (or an IT is flaky). Am investigating the failure.

paul-rogers commented:
Closing for now; too difficult to alter the ITs.
